### ISSUE TITLE
🐛 System.Exception: Unhandled MarshalAs attribute: FunctionPtr on field from iOS Bindings Library

### ISSUE THREAD

**Issue by phatblat** [xamarin/xamarin-macios#14871]
## TL;DR

The `UnmanagedType.FunctionPtr` enum value is not currently supported by the `MarshalAs()` attribute in Xamarin.iOS and .Mac.

## Description

I am working to create a Xamarin library for our cross-platform rust library and have gotten to a point where I am no longer getting helpful errors from MSBuild. Below are the two iOS bindings library projects involved:

- DittoFFI.iOS - bindings library, wrapping a native iOS framework built by rust/cargo
- DittoTransports.iOS - bindings library, wrapping a native iOS framework built by Xcode (depends on DittoFFI.iOS)

Both frameworks have only a single architecture for physical devices. Planning to convert both of these to XCFrameworks, but watching #10774 for resolution 👀

```
Non-fat file: DittoTransports.iOS/DittoTransports.framework/DittoTransports is architecture: arm64
Non-fat file: DittoFFI.iOS/dittoffi.framework/dittoffi is architecture: arm64
```

### Steps to Reproduce

1. Create an iOS C# Bindings library for a native library built by rust (DittoFFI.iOS). C# bindings code generated by [safer_ffi](https://github.com/getditto/safer_ffi)
2. Build the 1st bindings library, ensuring no build errors are produced
3. Create an iOS C# Bindings library for a native library (DittoTransports.iOS)
4. Generate c# bindings using Objective Sharpie, then fixup errors manually
5. Add a project reference to the 1st bindings library (DittoFFI.iOS)
6. BuIld the 2nd bindings library

### Expected Behavior

Building the DittoTransports.iOS project will succeed and it will be able to access the bound symbols from the referenced project, DittoFFI.iOS.

### Actual Behavior

Building the DittoTransports.iOS project fails citing problems with symbols defined in the referenced project, DittoFFI.iOS.

> System.Exception: Unhandled MarshalAs attribute: FunctionPtr on field Dittoffi.AwdlClientCallbacks_t.start_searching

The `AwdlClientCallbacks_t` struct compiles just fine in the DittoFFI.iOS project and another C# project for server platforms. Below is the struct definition for reference.

```csharp
[StructLayout(LayoutKind.Sequential, Size = 48)]
public unsafe struct AwdlClientCallbacks_t {
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public void_void_ptr_char_const_ptr_char_const_ptr_fptr start_searching;
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public void_void_ptr_fptr stop_searching;
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public void_void_ptr_char_const_ptr_fptr request_connect;
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public void_void_ptr_char_const_ptr_fptr request_disconnect;
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public int32_void_ptr_char_const_ptr_slice_ref_uint8_fptr send_data;
    [MarshalAs(UnmanagedType.FunctionPtr)]
    public int32_void_ptr_char_const_ptr_slice_mut_uint8_fptr read_data;
}
```

### Environment

<details>
  <summary>Version info</summary>

```
=== Visual Studio Community 2019 for Mac ===

Version 8.10.22 (build 11)
Installation UUID: 654d5689-aee8-4581-bcf6-d9189ee57026
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 6.18.0.23 (d16-6 / 088c73638)

	Package version: 612000174

=== Mono Framework MDK ===

Runtime:
	Mono 6.12.0.174 (2020-02/adf1bc4335d) (64-bit)
	Package version: 612000174

=== Roslyn (Language Service) ===

3.10.0-4.21269.26+029847714208ebe49668667c60ea5b0a294e0fcb

=== NuGet ===

Version: 5.9.0.7134

=== .NET SDK (x64) ===

SDK: /usr/local/share/dotnet/x64/sdk/6.0.202/Sdks
SDK Versions:
	7.0.202
	6.0.104
	5.0.407
	3.1.418
MSBuild SDKs: /Applications/Visual Studio.app/Contents/Resources/lib/monodevelop/bin/MSBuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/x64/dotnet
Runtime Versions:
	6.0.4
	5.0.16
	3.1.24

=== .NET Core 3.1 SDK ===

SDK: 3.1.418

=== .NET 5.0 SDK ===

SDK: 5.0.407

=== Xamarin.Profiler ===

Version: 1.8.0.19
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 13.3.1 (20103)
Build 13E500a

=== Xamarin Designer ===

Version: 16.11.0.60
Hash: 56f9b80b0
Branch: remotes/origin/d16-11
Build date: 2021-12-15 01:44:16 UTC

=== Xamarin.Mac ===

Version: 8.8.0.3 (Visual Studio Community)
Hash: eb7f8ed35
Branch: xcode13.3
Build date: 2022-03-30 03:16:09-0400

=== Xamarin.iOS ===

Version: 15.8.0.3 (Visual Studio Community)
Hash: eb7f8ed35
Branch: xcode13.3
Build date: 2022-03-30 03:16:09-0400

=== Xamarin.Android ===

Version: 12.2.4.3 (Visual Studio Community)
Commit: xamarin-android/d17-1/fc1125e
Android SDK: /Users/phatblat/Library/Android/sdk
	Supported Android versions:
		8.0 (API level 26)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 33.0.1
SDK Build Tools Version: 33.0.0 rc2

Build Information: 
Mono: adf1bc4
Java.Interop: xamarin/java.interop/d17-1@7ac33610
ProGuard: Guardsquare/proguard/v7.0.1@912d149
SQLite: xamarin/sqlite/3.37.1@0893b3b
Xamarin.Android Tools: xamarin/xamarin-android-tools/d17-1@db125a7

=== Eclipse Temurin JDK ===

Java SDK: Not Found

=== Android SDK Manager ===

Version: 16.10.0.13
Hash: 1b81df5
Branch: remotes/origin/d16-10
Build date: 2021-11-12 00:17:32 UTC

=== Android Device Manager ===

Version: 16.10.0.15
Hash: 89dcc0b
Branch: remotes/origin/d16-10
Build date: 2021-11-12 00:17:52 UTC

=== Build Information ===

Release ID: 810220011
Git revision: 16b49eb779f7def4523bbc8f4407628cf8dbb4b5
Build date: 2022-04-07 05:04:03-04
Build branch: release-8.10

=== Operating System ===

Mac OS X 12.3.1
Darwin 21.4.0 Darwin Kernel Version 21.4.0
    Fri Mar 18 00:46:32 PDT 2022
    root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 x86_64
```
</details>

### Build Logs

> CopyFilesToOutputDirectory:
>   DittoFFI.iOS -> /Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoFFI.iOS/bin/Debug/DittoFFI.iOS.dll
> Done Building Project "/Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoFFI.iOS/DittoFFI.iOS.csproj" (default targets).
> _GenerateBindings:
>   /Library/Frameworks/Xamarin.iOS.framework/Versions/Current/bin/bgen @/Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoTransports.iOS/obj/Debug/response-file.rsp
> BTOUCH : error BI0000: Unexpected error - Please file a bug report at https://github.com/xamarin/xamarin-macios/issues/new [/Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoTransports.iOS/DittoTransports.iOS.csproj]
>   System.Exception: Unhandled MarshalAs attribute: FunctionPtr on field Dittoffi.AwdlClientCallbacks_t.start_searching
>     at ObjCRuntime.Stret.GetValueTypeSize (System.Type original_type, System.Type type, System.Collections.Generic.List`1[T] field_types, System.Boolean is_64_bits, System.Int32& size, System.Int32& max_element_size, Generator generator) [0x00134] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at ObjCRuntime.Stret.GetValueTypeSize (System.Type type, System.Collections.Generic.List`1[T] fieldTypes, System.Boolean is_64_bits, Generator generator) [0x00070] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at ObjCRuntime.Stret.X86NeedStret (System.Type returnType, Generator generator) [0x0002b] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at ObjCRuntime.Stret.NeedStret (System.Type returnType, Generator generator) [0x00001] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at Generator.CheckNeedStret (System.Reflection.MethodInfo mi) [0x00008] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at Generator.DeclareInvoker (System.Reflection.MethodInfo mi) [0x0003e] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at Generator.Go () [0x00aab] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at BindingTouch.Main3 (System.String[] args) [0x01279] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at BindingTouch.Main2 (System.String[] args) [0x00007] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at BindingTouch.Main (System.String[] args) [0x00002] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at System.Environment.get_StackTrace () [0x00000] in /Users/builder/jenkins/workspace/build-package-osx-mono/2020-02/external/bockbuild/builds/mono-x64/mcs/class/corlib/System/Environment.cs:316
>     at ErrorHelper.ShowInternal (System.Exception e) [0x00000] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at ErrorHelper.Show (System.Exception e, System.Boolean rethrow_errors) [0x00000] in <fb75fb38924e41d9a5878f90298e1b3f>:0
>     at BindingTouch.Main (System.String[] args) [0x00000] in <fb75fb38924e41d9a5878f90298e1b3f>:0
> Done Building Project "/Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoTransports.iOS/DittoTransports.iOS.csproj" (default targets) -- FAILED.
> 
> Build FAILED.
>
> (_GenerateBindings target) ->
>  BTOUCH : error BI0000: Unexpected error - Please file a bug report at https://github.com/xamarin/xamarin-macios/issues/new [/Users/phatblat/dev/ditto/ditto/dotnet/Xamarin/DittoTransports.iOS/DittoTransports.iOS.csproj]

**Comment by phatblat** [comment]
The main question I have is: can an iOS bindings library project depend on another iOS bindings library project? Sure seems like the tools don't support it.

**Comment by chamons** [comment]
The exception in question is found here:

https://github.com/xamarin/xamarin-macios/blob/main/src/ObjCRuntime/Stret.cs#L337

which suggests we are trying to get the size of a value type, and that the type isn't one we support.

The fact that you define it as a struct might be why we're falling in here, and one or more of those types might not be supported.

Having a sample project that reproduces this would be very helpful, else at best we are guessing if our fix solves your issue. At minimum a copy of the generated code would allow us to at least see the full type definitions. 

@rolfbjarne - This sounds like trampoline code, could you review to see if there is something I missed?

**Comment by danielhenrymantilla** [comment]
Hey, author of [`::safer-ffi`] here. Given the error message and what you've pointed out, @chamons, it does indeed look like a special case of returning a "composite data structure" / a `struct` which has a `delegate` field with a `[MarshalAs(UnmanagedType.FunctionPtr)]` directive on it.

I think a small repro of such API would be the following C snippet:

```c
// ...

typedef struct cb {
    void * data;
    void (*cb)(void * data);
} cb_t;

cb_t get_cb (void) {
    // ...
```

  - [Godbolt](https://c.godbolt.org/z/1TG94rvcx)

___

With the equivalent definition, in Rust (which can be compiled down to a static or dynamic library with this C API), and then using [`::safer-ffi`] to generate some equivalent C# headers directly, we get:

[`::safer-ffi`]: https://getditto.github.io/safer_ffi/

```csharp
/*! \file */
/*******************************************
 *                                         *
 *  File auto-generated by `::safer_ffi`.  *
 *                                         *
 *  Do not manually edit this file.        *
 *                                         *
 *******************************************/

// ...

namespace <projectname> {
using System;
using System.Runtime.InteropServices;

public unsafe partial class Ffi {
    private const string RustLib = "<path to lib>";
}

[UnmanagedFunctionPointer(CallingConvention.Winapi)]
public unsafe /* static */ delegate
    void
    void_void_ptr_fptr (
         void *    _0);

[StructLayout(LayoutKind.Sequential, Size = 16)]
public unsafe struct Cb_t {
    void * data;

    [MarshalAs(UnmanagedType.FunctionPtr)]
    public void_void_ptr_fptr cb;
}

public unsafe partial class Ffi {
    [DllImport(RustLib, ExactSpelling = true)] public static unsafe extern
    Cb_t get_cb ();
}

} /* <projectname> */
```

so for the sake of the reproduction, it would suffice to compile that C code above as a C library, and use this "C# header" to be able to call into it.

**Comment by rolfbjarne** [comment]
@chamons is right, we need to add support for `UnmanagedType.FunctionPtr` here:

https://github.com/xamarin/xamarin-macios/blob/20212e6e26c49ee83bb7febc17b816a93d079845/src/ObjCRuntime/Stret.cs#L312-L338

This seems simple enough to add (although there might be other issues once this is fixed, so we'd probably have to create some test code to make sure the scenario actually works).

**Comment by rolfbjarne** [comment]
I'm not sure it's worth supporting MarshalAs.FunctionPtr:

1. It's the default marshalling behavior: https://learn.microsoft.com/en-us/dotnet/framework/interop/default-marshalling-behavior#default-marshalling-for-delegates.
2. In .NET, the recommended approach is to use function pointers instead (which are faster at runtime, and has a few other benefits as well): https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers

So unless there's a compelling argument otherwise, we'll close this issue.
